From 1b328050685d288cb185847977de4de4a1c7ea84 Mon Sep 17 00:00:00 2001
From: Sean Busbey <busbey@apache.org>
Date: Mon, 16 Oct 2017 16:11:39 -0500
Subject: [PATCH 421/438] HBASE-19020 HBase Rest test for xml parsing external
 entities should not rely on implementation of java
 XML APIs.

Signed-off-by: Chia-Ping Tsai <chia7712@gmail.com>
(cherry picked from commit b4cf88b7697cf6bf48647f4387614f681d849751)

Change-Id: I2f31a0f51830c3d863b7ded30769851bf4926d59
Author: Sean Busbey
Reason: Test
Ref: CDH-61626
---
 .../hadoop/hbase/rest/client/TestXmlParsing.java   |   10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

diff --git a/hbase-rest/src/test/java/org/apache/hadoop/hbase/rest/client/TestXmlParsing.java b/hbase-rest/src/test/java/org/apache/hadoop/hbase/rest/client/TestXmlParsing.java
index 56dc05e..0b94e9e 100644
--- a/hbase-rest/src/test/java/org/apache/hadoop/hbase/rest/client/TestXmlParsing.java
+++ b/hbase-rest/src/test/java/org/apache/hadoop/hbase/rest/client/TestXmlParsing.java
@@ -23,7 +23,10 @@ import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.when;
 
 import java.io.IOException;
+import javax.xml.bind.UnmarshalException;
 
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
 import org.apache.hadoop.hbase.HBaseConfiguration;
 import org.apache.hadoop.hbase.rest.Constants;
 import org.apache.hadoop.hbase.rest.model.StorageClusterVersionModel;
@@ -37,6 +40,7 @@ import org.junit.experimental.categories.Category;
  */
 @Category(SmallTests.class)
 public class TestXmlParsing {
+  private static final Log LOG = LogFactory.getLog(TestXmlParsing.class);
 
   @Test
   public void testParsingClusterVersion() throws Exception {
@@ -68,8 +72,12 @@ public class TestXmlParsing {
       admin.getClusterVersion();
       fail("Expected getClusterVersion() to throw an exception");
     } catch (IOException e) {
+      assertEquals("Cause of exception ought to be a failure to parse the stream due to our " +
+          "invalid external entity. Make sure this isn't just a false positive due to " +
+          "implementation. see HBASE-19020.", UnmarshalException.class, e.getCause().getClass());
       final String exceptionText = StringUtils.stringifyException(e);
-      final String expectedText = "The entity \"xee\" was referenced, but not declared.";
+      final String expectedText = "\"xee\"";
+      LOG.debug("exception text: '" + exceptionText + "'", e);
       assertTrue("Exception does not contain expected text", exceptionText.contains(expectedText));
     }
   }
-- 
1.7.9.5

